### DISCUSSION TITLE
How to test in FastAPI that client calls correct method of endpoint?

### DISCUSSION THREAD

**Discussion by FyZzyss** [fastapi/fastapi#7223] [Questions] [answered]
I want to create test with `pytest` which check that user calls correct method of endpoint.
How I can do this?
P.S. I tried to do it on my own, but it doesn't work. My code will be below.
**My test:**
```
@pytest.mark.unit
def test_get_best_authors(mocker: MockFixture):
    test = mocker.spy(main, 'best_authors')  
    response = client.get('/luchshie-avtori')
    assert response.status_code == 200
    test.assert_called_once()
```
**My main.py with endpoints:**
```
@app.get("/luchshie-avtori")
async def best_authors():
    return requests.get('', params={'return': 'json'}).json()
```

**Comment by la-mar** [comment]
Here's a working example that extends the code you provided.  httpx is an async alternative to requests and has a nice mechanism for testing your FastAPI app via its AsyncClient (see the pytest fixture).

I changed your endpoint to call what I assume is a third party API using httpx instead of requests since you defined the endpoint as async.  You can revert it to use requests, but be sure to define your endpoint as a synchronous function instead of async (only "def" instead of "async def").

```python
#  your app
import httpx
from fastapi import FastAPI, status


app = FastAPI()


@app.get("/luchshie-avtori")
async def best_authors():
    return httpx.get(
        "https://jsonplaceholder.typicode.com/users", params={"return": "json"}
    ).json()


# your test file
import pytest


@pytest.fixture
async def client(bind):
    """ Create a test client connected to the main app instance. """

    # http://testserver is an httpx "magic" url that tells the client to query the
    # given app instance.
    async with httpx.AsyncClient(app=app, base_url="http://testserver") as client:
        yield client


@pytest.mark.asyncio
async def test_get_best_authors(client):

    # call the endpoint using the client
    response = await client.get("/luchshie-avtori")

    # test response is okay
    assert response.status_code == status.HTTP_200_OK

    data = response.json()
    # make other assertions on data
    assert data is not None

```

**Comment by tiangolo** [comment]
Thanks for the help @la-mar ! :coffee: 

I assume you solved your issue @FyZzyss, so thanks for closing it :+1:

**Comment by YuriiMotov** [comment, marked as answer]
You can't mock the endpoint function this way.

I'll try to explain why, feel free to skip this part:
The reason is that at the moment your test is run, FastAPI already too the value of `best_authors` object (functions are also objects) and stored it internally (in APIRoute object). To run this endpoint it uses that internal pointer to that object (function). So, mocking the `best_authors` object will not take effect.

In your case I would mock `request.get` and assert that it was called with right parameters
